src/lib.rs

```diff
@@ -1,350 +1,440 @@
-use std::collections::HashMap;
+use std::collections::BTreeMap;
 
-/// Compute the gini impurity of a dataset. 
-/// 
+/// Compute the gini impurity of a dataset.
+///
 /// Returns a float, 0 representing a perfectly pure dataset. Normal distribution: ~0.33
-/// 
+///
 /// By default, any empty dataset will return a gini of 1.0. This may be unexpected behaviour.
 /// ```
 /// use parsnip::gini;
-/// assert_eq!(gini(&vec![0, 0, 0, 1]), 0.375);
+/// assert_eq!(gini(&vec![0_usize, 0, 0, 1]), 0.375);
 /// ```
-pub fn gini(data: &[u64]) -> f32 {
+pub fn gini<T>(data: &[T]) -> f32
+where
+    T: Eq,
+    T: Ord,
+{
     if data.len() == 0 {
         return 1.0;
-    } 
+    }
     fn p_squared(count: usize, len: f32) -> f32 {
         let p = count as f32 / len;
         p * p
     }
     let len = data.len() as f32;
-    let mut count = HashMap::new();
-    for &value in data {
+    let mut count = BTreeMap::new();
+    for value in data.iter() {
         *count.entry(value).or_insert(0) += 1;
     }
     let counts: Vec<usize> = count.into_iter().map(|(_, c)| c).collect();
-    let indiv : Vec<f32> = counts.iter().map(|x| p_squared(*x, len)).collect();
-    let sum : f32 = indiv.iter().sum();
-    1.0 - sum
+    let indiv: Vec<f32> = counts.iter().map(|x| p_squared(*x, len)).collect();
```

Comment:
Again, I don't think `counts` or `indiv` have to be collected into vectors here. It's inefficient to store intermediate results into vectors without using that content more than once.

Reply:
Also cleaned up in the latest commit where possible—there's been a couple that if I kept chaining gave type errors, but minimized as much as possible 👍